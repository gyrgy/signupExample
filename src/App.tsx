import React from 'react';
import { Container } from '@material-ui/core';

const App = () => {
  return (
    <Container>
      <div>App</div>
    </Container>
  );
};

export default App;
