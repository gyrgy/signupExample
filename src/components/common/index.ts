export { default as ButtonPrimary } from 'src/components/common/buttonPrimary';
export { default as Checkbox } from 'src/components/common/checkbox';
export { default as Content } from 'src/components/common/content';
export { default as Fade } from 'src/components/common/fade';
export { default as Form } from 'src/components/common/form';
export { default as InputField } from 'src/components/common/inputField';
export { default as TabPanel } from 'src/components/common/tabPanel';