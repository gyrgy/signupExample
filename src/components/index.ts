export { default as SignUpDone } from 'src/components/signUpDone';
export { default as SignUpPrivacyForm } from 'src/components/signUpPrivacyForm';
export { default as SignUpTabMenu } from 'src/components/signUpTabMenu';
export { default as SignUpUserForm } from 'src/components/signUpUserForm';
