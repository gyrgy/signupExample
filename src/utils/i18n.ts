import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import { translationEn } from 'src/locales/en';

i18n.use(initReactI18next).init({
  resources: {
    en: {
      translations: {
        ...translationEn,
      },
    },
  },
  fallbackLng: 'en',
  interpolation: {
    escapeValue: true,
  },
  ns: ['translations'],
  defaultNS: 'translations',
});

export const t = (key: string, props?: object): string => i18n.t(key, props);
