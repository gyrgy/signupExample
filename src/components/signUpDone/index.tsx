import React from 'react';
import { Alert } from '@material-ui/lab';
import { signUpDoneTestID } from 'src/constants/testIDs';
import { t } from 'src/utils/i18n';

interface SignUpDoneProps {
  testID?: string;
}

const SignUpDone = (props: SignUpDoneProps) => {
  const { testID } = props;
  return (
    <div data-testid={testID || signUpDoneTestID}>
      <Alert severity="success">
        {t('doneForm')}
      </Alert>
    </div>
  );
};

export default SignUpDone;