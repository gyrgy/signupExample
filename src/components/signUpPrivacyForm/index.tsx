import React, { ChangeEvent } from 'react';
import { ButtonPrimary, Checkbox, Form } from 'src/components/common';
import { PrivacyFormCheckboxes } from 'src/constants/enums';
import { t } from 'src/utils/i18n';
import { privacyFormTestID, privacyFormTrayCheckboxTestID, privacyFormOtherCheckboxTestID } from 'src/constants/testIDs';

interface PrivacyFormProps {
  trayProductUpdatesChecked: boolean;
  otherProductsUpdatesChecked: boolean;
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
  submitForm: () => void;
}

const PrivacyForm = (props: PrivacyFormProps) => {
  const {
    trayProductUpdatesChecked,
    otherProductsUpdatesChecked,
    onChange,
    submitForm
  } = props;

  return (
    <Form testID={privacyFormTestID}>
      <Checkbox
        name={PrivacyFormCheckboxes.TrayProductUpdatesCheckboxName}
        label={t('privacyForm.trayProductUpdatesCheckboxLabel')}
        checked={trayProductUpdatesChecked}
        onChange={onChange}
        testID={privacyFormTrayCheckboxTestID}
      />
      <Checkbox
        name={PrivacyFormCheckboxes.OtherProductsUpdatesCheckboxName}
        label={t('privacyForm.otherProductsUpdatesCheckboxLabel')}
        checked={otherProductsUpdatesChecked}
        onChange={onChange}
        testID={privacyFormOtherCheckboxTestID}
      />
      <ButtonPrimary label="submit" onClick={submitForm} />
    </Form>
  );
};

export default PrivacyForm;