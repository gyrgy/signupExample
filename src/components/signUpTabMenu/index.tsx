import React, { ChangeEvent } from 'react';
import { AppBar, Tab, Tabs } from '@material-ui/core';
import { t } from 'src/utils/i18n';
import { tabMenuTestID, tabMenuUserTabTestID, tabMenuPrivacyTabTestID, tabMenuDoneTabTestID } from 'src/constants/testIDs';

interface TabMenuProps {
  selectedIndex: number;
  onChange: (index: number) => void;
  testID?: string;
}

const SignUpTabMenu = (props: TabMenuProps) => {
  const { selectedIndex, onChange, testID } = props;

  const handleTabChange = (event: ChangeEvent<{}>, newValue: number) => {
    onChange(newValue);
  };

  return (
    <AppBar position="static" color="default">
      <Tabs
        value={selectedIndex}
        onChange={handleTabChange}
        indicatorColor="primary"
        textColor="primary"
        variant="fullWidth"
        aria-label="full width tabs"
        data-testid={testID || tabMenuTestID}
      >
        <Tab label={t('tabsData.user')} disabled={selectedIndex !== 0} data-testid={tabMenuUserTabTestID} />
        <Tab label={t('tabsData.privacy')} disabled={selectedIndex !== 1} data-testid={tabMenuPrivacyTabTestID} />
        <Tab label={t('tabsData.done')} disabled={selectedIndex !== 2} data-testid={tabMenuDoneTabTestID} />
      </Tabs>
    </AppBar>
  );
};

export default SignUpTabMenu;
