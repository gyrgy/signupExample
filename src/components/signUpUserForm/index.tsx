import React, { ChangeEvent, FocusEvent } from 'react';
import { Form, InputField, ButtonPrimary } from 'src/components/common';
import { UserFormFieldNames } from 'src/constants/enums';
import { t } from 'src/utils/i18n';
import {
  userFormTestID,
  userFormNameFieldTestID,
  userFormRoleFieldTestID,
  userFormEmailFieldTestID,
  userFormPasswordFieldTestID
} from 'src/constants/testIDs';

interface SignUpUserFromProps {
  userNameFieldValue: string;
  userNameFieldError: boolean;
  userRoleFieldValue: string;
  userRoleFieldError: boolean;
  userEmailFieldValue: string;
  userEmailFieldError: boolean;
  userPasswordFieldValue: string;
  userPasswordFieldError: boolean;
  handleChange: (event: ChangeEvent<HTMLInputElement>) => void;
  handleBlur:
  (event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => void;
  handleSubmit: () => void;
  testID?: string;
}

const SignUpUserFrom = (props: SignUpUserFromProps) => {
  const {
    userNameFieldValue,
    userNameFieldError,
    userRoleFieldValue,
    userRoleFieldError,
    userEmailFieldValue,
    userEmailFieldError,
    userPasswordFieldValue,
    userPasswordFieldError,
    handleChange,
    handleBlur,
    handleSubmit,
  } = props;

  return (
    <Form testID={userFormTestID}>
      <InputField
        type="text"
        label={t('userForm.userNameFieldLabel')}
        name={UserFormFieldNames.Name}
        onChange={handleChange}
        onBlur={handleBlur}
        value={userNameFieldValue}
        required
        error={userNameFieldError}
        errorHelperText={userNameFieldError ? t('userForm.userNameFieldErrorHelperText') : ''}
        testID={userFormNameFieldTestID}
      />
      <InputField
        type="text"
        label={t('userForm.userRoleFieldLabel')}
        name={UserFormFieldNames.Role}
        onChange={handleChange}
        onBlur={handleBlur}
        value={userRoleFieldValue}
        error={userRoleFieldError}
        errorHelperText={userRoleFieldError ? t('userForm.userRoleFieldErrorHelperText') : ''}
        testID={userFormRoleFieldTestID}
      />
      <InputField
        type="text"
        label={t('userForm.userEmailFieldLabel')}
        name={UserFormFieldNames.Email}
        onChange={handleChange}
        onBlur={handleBlur}
        value={userEmailFieldValue}
        required
        error={userEmailFieldError}
        errorHelperText={userEmailFieldError ? t('userForm.userEmailFieldErrorHelperText') : ''}
        testID={userFormEmailFieldTestID}
      />
      <InputField
        type="password"
        label={t('userForm.userPasswordFieldLabel')}
        name={UserFormFieldNames.Password}
        onChange={handleChange}
        onBlur={handleBlur}
        value={userPasswordFieldValue}
        required
        error={userPasswordFieldError}
        errorHelperText={userPasswordFieldError ? t('userForm.userPasswordFieldErrorHelperText') : ''}
        testID={userFormPasswordFieldTestID}
      />
      <ButtonPrimary label="submit" onClick={handleSubmit} />
    </Form>
  );
};

export default SignUpUserFrom;
