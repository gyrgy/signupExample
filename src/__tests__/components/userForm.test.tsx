import React from 'react';
import {
  render,
  RenderResult,
  cleanup,
  fireEvent
} from '@testing-library/react';
import { SignUpUserForm } from 'src/components';
import {
  userFormTestID,
  userFormNameFieldTestID,
  userFormRoleFieldTestID,
  userFormEmailFieldTestID,
  userFormPasswordFieldTestID,
  primaryButtonTestID,
} from 'src/constants/testIDs';
import { t } from 'src/utils/i18n';

describe('test tab panel component', () => {
  const userNameFieldValue = 'userNameFieldValue';
  let userNameFieldError = false;
  const userRoleFieldValue = 'userRoleFieldValue';
  let userRoleFieldError = false;
  const userEmailFieldValue = 'userEmailFieldValue';
  let userEmailFieldError = false;
  const userPasswordFieldValue = 'userPasswordFieldValue';
  let userPasswordFieldError = false;
  const spyOnChangeHandler = jest.fn();
  const spyOnBlurHandler = jest.fn();
  const spyOnSubmitHandler = jest.fn();

  let form: RenderResult;

  beforeEach(() => {
    form = render(
      <SignUpUserForm
        userNameFieldValue={userNameFieldValue}
        userNameFieldError={userNameFieldError}
        userRoleFieldValue={userRoleFieldValue}
        userRoleFieldError={userRoleFieldError}
        userEmailFieldValue={userEmailFieldValue}
        userEmailFieldError={userEmailFieldError}
        userPasswordFieldValue={userPasswordFieldValue}
        userPasswordFieldError={userPasswordFieldError}
        handleChange={spyOnChangeHandler}
        handleBlur={spyOnBlurHandler}
        handleSubmit={spyOnSubmitHandler}
      />
    );
  });

  afterEach(cleanup);

  it('renders user form snd children', () => {
    const { getByTestId } = form;
    const userForm = getByTestId(userFormTestID);
    const nameField = getByTestId(userFormNameFieldTestID);
    const roleField = getByTestId(userFormRoleFieldTestID);
    const emailField = getByTestId(userFormEmailFieldTestID);
    const passwordField = getByTestId(userFormPasswordFieldTestID);
    const submitButton = getByTestId(primaryButtonTestID);

    expect(userForm).toBeInTheDocument();
    expect(nameField).toBeInTheDocument();
    expect(roleField).toBeInTheDocument();
    expect(emailField).toBeInTheDocument();
    expect(passwordField).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });

  it('tests callbacks', () => {
    const { getByTestId } = form;

    const inputField = getByTestId(userFormNameFieldTestID)
      .querySelector('input[type="text"]') as HTMLInputElement;
    fireEvent.change(inputField, { target: { value: 'test value' } });
    expect(spyOnChangeHandler).toHaveBeenCalledTimes(1);
    fireEvent.blur(inputField);
    expect(spyOnBlurHandler).toHaveBeenCalledTimes(1);

    const button = getByTestId(primaryButtonTestID);
    fireEvent.click(button);
    expect(spyOnSubmitHandler).toHaveBeenCalledTimes(1);
  });

  it('tests error state', () => {
    userNameFieldError = true;
    userRoleFieldError = true;
    userEmailFieldError = true;
    userPasswordFieldError = true;

    const { rerender, getByText } = form;
    rerender(
      <SignUpUserForm
        userNameFieldValue={userNameFieldValue}
        userNameFieldError={userNameFieldError}
        userRoleFieldValue={userRoleFieldValue}
        userRoleFieldError={userRoleFieldError}
        userEmailFieldValue={userEmailFieldValue}
        userEmailFieldError={userEmailFieldError}
        userPasswordFieldValue={userPasswordFieldValue}
        userPasswordFieldError={userPasswordFieldError}
        handleChange={spyOnChangeHandler}
        handleBlur={spyOnBlurHandler}
        handleSubmit={spyOnSubmitHandler}
      />
    );

    const nameErrorHelper =
      getByText(t('userForm.userNameFieldErrorHelperText'));
    const roleErrorHelper =
      getByText(t('userForm.userRoleFieldErrorHelperText'));
    const emailErrorHelper =
      getByText(t('userForm.userEmailFieldErrorHelperText'));
    const passwordErrorHelper =
      getByText(t('userForm.userPasswordFieldErrorHelperText'));

    expect(nameErrorHelper).toBeInTheDocument();
    expect(roleErrorHelper).toBeInTheDocument();
    expect(emailErrorHelper).toBeInTheDocument();
    expect(passwordErrorHelper).toBeInTheDocument();
  });

  it('tests if matches with snapshot', () => {
    const { asFragment } = form;
    expect(asFragment()).toMatchSnapshot();
  });
});