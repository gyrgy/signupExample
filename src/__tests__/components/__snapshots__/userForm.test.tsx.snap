// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test tab panel component tests if matches with snapshot 1`] = `
<DocumentFragment>
  <form
    autocomplete="off"
    class="makeStyles-root-324"
    data-testid="components/signUpUserForm"
    novalidate=""
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      data-testid="components/signUpUserForm/nameField"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-error Mui-error MuiFormLabel-filled Mui-required Mui-required"
        data-shrink="true"
      >
        name
        <span
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error Mui-error"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-invalid="true"
          class="MuiInputBase-input MuiOutlinedInput-input"
          name="name"
          required=""
          type="text"
          value="userNameFieldValue"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-383 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-385 PrivateNotchedOutline-legendNotched-386"
          >
            <span>
              name *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained Mui-error MuiFormHelperText-filled Mui-required"
      >
        This field is required and can contain only letters.
      </p>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      data-testid="components/signUpUserForm/roleField"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-error Mui-error MuiFormLabel-filled"
        data-shrink="true"
      >
        role
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-invalid="true"
          class="MuiInputBase-input MuiOutlinedInput-input"
          name="role"
          type="text"
          value="userRoleFieldValue"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-383 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-385 PrivateNotchedOutline-legendNotched-386"
          >
            <span>
              role
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained Mui-error MuiFormHelperText-filled"
      >
        This field can contain only letters or you can leave it blank
      </p>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      data-testid="components/signUpUserForm/emailField"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-error Mui-error MuiFormLabel-filled Mui-required Mui-required"
        data-shrink="true"
      >
        email
        <span
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error Mui-error"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-invalid="true"
          class="MuiInputBase-input MuiOutlinedInput-input"
          name="email"
          required=""
          type="text"
          value="userEmailFieldValue"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-383 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-385 PrivateNotchedOutline-legendNotched-386"
          >
            <span>
              email *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained Mui-error MuiFormHelperText-filled Mui-required"
      >
        Please provide a valid email address.
      </p>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      data-testid="components/signUpUserForm/passwordField"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-error Mui-error MuiFormLabel-filled Mui-required Mui-required"
        data-shrink="true"
      >
        password
        <span
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error Mui-error"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-invalid="true"
          class="MuiInputBase-input MuiOutlinedInput-input"
          name="password"
          required=""
          type="password"
          value="userPasswordFieldValue"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-383 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-385 PrivateNotchedOutline-legendNotched-386"
          >
            <span>
              password *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained Mui-error MuiFormHelperText-filled Mui-required"
      >
        The password has to be 9 characters long, has to contain at least one number, one lowercase and on uppercase character.
      </p>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      data-testid="components/constants/buttonPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        submit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
